checks/class-skip-links.php

```diff
@@ -0,0 +1,150 @@
+ */
+class Skip_Links_Check implements themecheck {
+	/**
+	 * Error messages, warnings and info notices.
+	 *
+	 * @var array $error
+	 */
+	protected $error = array();
+
+	/**
+	 * Returns true if the theme is a block theme.
+	 *
+	 * @var array $is_block_theme
+	 */
+	protected $is_block_theme = false;
+
+	/**
+	 * The WP_Theme instance being checked.
+	 *
+	 * @var WP_Theme $wp_theme
+	 */
+	protected $wp_theme = false;
+
+	function set_context( $data ) {
+		if ( isset( $data['theme'] ) ) {
+			$this->wp_theme       = $data['theme'];
+			$this->is_block_theme = wp_is_block_theme();
+		}
+	}
+
+	/**
+	 * Check that return true for good/okay/acceptable, false for bad/not-okay/unacceptable.
+	 *
+	 * @param array $php_files File paths and content for PHP files.
+	 * @param array $css_files File paths and content for CSS files.
+	 * @param array $other_files Folder names, file paths and content for other files.
+	 */
+	public function check( $php_files, $css_files, $other_files ) {
+
+		$info                       = '';
+		$templates_without_main_tag = array();
+
+		$directory = 'templates'; // Path to the folder containing HTML files
+		$theme_dir = $this->wp_theme->get_stylesheet_directory();
+
+		// Get all HTML files in the directory
+		$files = glob( $theme_dir . '/' . $directory . '/*.html' );
+
+		foreach ( $files as $file ) {
+			$contents     = file_get_contents( $file );
+			$has_main_tag = strpos( $contents, '<main' ) !== false;
+			$file_name    = basename( $file );
+
+			if ( ! $has_main_tag ) {
+				$pattern_slugs = $this->template_has_patterns( $contents );
+				if ( $pattern_slugs ) {
+					foreach ( $pattern_slugs as $slug ) {
+						$has_main_tag = $this->pattern_has_tag( $slug );
+						if ( ! $has_main_tag ) {
+							$templates_without_main_tag[] = $file_name;
+						}
+					}
+				} else {
+					$templates_without_main_tag[] = $file_name;
+				}
+			}
+		}
+
+		$info = implode( ', ', $templates_without_main_tag );
+
+		if ( $info !== '' ) {
+			$this->error[] = sprintf(
+				'<span class="tc-lead tc-required">%s</span> %s ',
+				__( 'REQUIRED', 'theme-check' ),
+				sprintf(
+					__( 'Skip links are missing from the following templates: %s Please make sure the templates have a &lt;main&gt; tag.', 'theme-check' ),
+					$info
+				)
+			);
+		}
+
+		return true;
+	}
+
+	function template_has_patterns( $contents ) {
+		$pattern = '/<!-- wp:pattern \{"slug":"([^"]+)"\} \/-->/';
+		if ( preg_match_all( $pattern, $contents, $matches ) ) {
+			$slugs = $matches[1];
+			return $slugs;
+		} else {
+			return false;
+		}
+	}
+
+	function pattern_has_tag( $slug ) {
+		$directory = 'patterns';
+		$theme_dir = $this->wp_theme->get_stylesheet_directory();
+
+		if ( ! is_dir( $theme_dir . '/' . $directory ) ) {
+			$directory = 'block-patterns';
+		}
+		if ( ! is_dir( $theme_dir . '/' . $directory ) ) {
+			return false;
+		}
+
+		$files = glob( $theme_dir . '/' . $directory . '/*.php' );
```

Comment:
Same as in https://github.com/WordPress/theme-check/pull/457#discussion_r1761338458

Reply:
It looks like `$php_files` doesn't store the contents of the comment in the pattern, so I have no way of checking if the slug of the pattern is the one I'm trying to match

Reply:
seems like that's done here so I will leave this part of the code as I implemented it: https://github.com/WordPress/theme-check/blob/b2b4cec4db511691827068efb3d4010911c0d66d/checkbase.php#L61